src/per/qspi.cpp

```diff
@@ -0,0 +1,1008 @@
+#include "per/qspi.h"
+#include "stm32h7xx_hal.h"
+#include "dev/flash_IS25LP080D.h"
+#include "dev/flash_IS25LP064A.h"
+extern "C"
+{
+#include "util/hal_map.h"
+}
+
+// TODO: Add handling for alternate device types,
+//        This will be a thing much sooner than anticipated
+//        due to upgrading the RAM size for the new 4MB chip.
+// TODO: AutopollingMemReady only works for 1-Line, not 4-Line
+
+#define SET_MODE(mode)                              \
+    if(mode_ != mode)                               \
+    {                                               \
+        mode_ = mode;                               \
+        if(Init(config_) != QSPIHandle::Result::OK) \
+        {                                           \
+            mode_ = Mode::MEMORY_MAPPED;            \
+            Init(config_);                          \
+            return QSPIHandle::Result::ERR;         \
+        }                                           \
+    }
```

Comment:
why is this a macro? This can/should be an Impl function.

Having it be a functions should also help with the notes on "code size" where you're using gotos and jumping to this macro. Without it being a function, every instance is forced into the flash size, and can't easily be optimized.